rasa/core/policies/ensemble.py

```diff
@@ -91,590 +218,315 @@ def _check_for_important_policies(self) -> None:
-            diagnostic_data=best_prediction.diagnostic_data,
-            hide_rule_turn=best_prediction.hide_rule_turn,
-        )
-
-    def _best_policy_prediction(
-        self,
-        tracker: DialogueStateTracker,
-        domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
-    ) -> PolicyPrediction:
-        """Finds the best policy prediction.
+        Note that this comparison is *not* symmetric if the priorities are allowed to
+        coincide (i.e. if we cannot distinguish two predictions using 1.-4., then
+        the first prediction is considered to be "better").
 
         Args:
-            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
-            domain: the :class:`rasa.shared.core.domain.Domain`
-            interpreter: Interpreter which may be used by the policies to create
-                additional features.
-
+          predictions: all predictions made by an ensemble of policies
         Returns:
-            The winning policy prediction.
+          index of the best prediction
         """
-        # find rejected action before running the policies
-        # because some of them might add events
-        rejected_action_name = None
-        last_action_event = next(
-            (
-                event
-                for event in reversed(tracker.events)
-                if isinstance(event, (ActionExecutionRejected, ActionExecuted))
-            ),
-            None,
+        does_not_contain_no_user_predictions = all(
+            not prediction.is_no_user_prediction for prediction in predictions
         )
 
-        if len(tracker.events) > 0 and isinstance(
-            last_action_event, ActionExecutionRejected
-        ):
-            rejected_action_name = last_action_event.action_name
-
-        predictions = {
-            f"policy_{i}_{type(p).__name__}": p.predict_action_probabilities(
-                tracker, domain, interpreter
+        def criteria(prediction: PolicyPrediction2) -> Tuple[bool, bool, float, int]:
+            """Extracts statistics (larger is better) ordered by importance."""
+            return (
+                prediction.is_no_user_prediction,
+                does_not_contain_no_user_predictions
+                and prediction.is_end_to_end_prediction,
+                prediction.max_probability,
+                prediction.policy_priority,
             )
-            for i, p in enumerate(self.policies)
-        }
 
-        if rejected_action_name:
-            logger.debug(
-                f"Execution of '{rejected_action_name}' was rejected. "
-                f"Setting its confidence to 0.0 in all predictions."
-            )
-            for prediction in predictions.values():
-                prediction.probabilities[
-                    domain.index_for_action(rejected_action_name)
-                ] = 0.0
-
-        return self._pick_best_policy(predictions)
+        return max(
```

Comment:
somewhat neat but not sure if more readable 🤔

Reply:
just the max or the whole function? I can pull it out and name it argmax